File: packages/python/elizaos/features/advanced_capabilities/experience/evaluators/experience_evaluator.py
from __future__ import annotations

import json
import logging
import re
from collections.abc import Awaitable, Callable
from typing import TYPE_CHECKING

from elizaos.features.advanced_capabilities.experience.service import EXPERIENCE_SERVICE_TYPE
from elizaos.features.advanced_capabilities.experience.types import ExperienceType, OutcomeType
from elizaos.types import ActionResult, Evaluator, HandlerOptions, ModelType

if TYPE_CHECKING:
    from elizaos.types import Content, IAgentRuntime, Memory, State

logger = logging.getLogger("elizaos.experience")

# Domain detection keywords (matches TypeScript implementation)
_DOMAIN_KEYWORDS: dict[str, list[str]] = {
    "shell": ["command", "terminal", "bash", "shell", "execute", "script", "cli"],
    "coding": [
        "code",
        "function",
        "variable",
        "syntax",
        "programming",
        "debug",
        "typescript",
        "javascript",
    ],
    "system": [
        "file",
        "directory",
        "process",
        "memory",
        "cpu",
        "system",
        "install",
        "package",
    ],
    "network": ["http", "api", "request", "response", "url", "network", "fetch", "curl"],
    "data": ["json", "csv", "database", "query", "data", "sql", "table"],
    "ai": ["model", "llm", "embedding", "prompt", "token", "inference"],
}

_EXPERIENCE_TYPE_MAP: dict[str, ExperienceType] = {
    "DISCOVERY": ExperienceType.DISCOVERY,
    "CORRECTION": ExperienceType.CORRECTION,
    "SUCCESS": ExperienceType.SUCCESS,
    "LEARNING": ExperienceType.LEARNING,
}

# Prompt template for extracting experiences from conversation
_EXTRACT_EXPERIENCES_TEMPLATE = """Analyze the following conversation context and extract any novel learning experiences.

Conversation context:
{{conversation_context}}

Existing experiences (to avoid duplicates):
{{existing_experiences}}

Extract experiences as a JSON array. Each experience should have:
- type: one of DISCOVERY, CORRECTION, SUCCESS, LEARNING
- learning: what was learned (concise, actionable)
- context: what was happening when this was learned
- confidence: 0-1 how confident in this learning
- reasoning: brief explanation of why this is notable

Return ONLY a JSON array, no other text. Return [] if no novel experiences found.
"""


def _detect_domain(text: str) -> str:
    """Detect the domain of an experience from its text content."""
    lower_text = text.lower()
    for domain, keywords in _DOMAIN_KEYWORDS.items():
        if any(keyword in lower_text for keyword in keywords):
            return domain
    return "general"


def _sanitize_context(text: str) -> str:
    """Remove user-specific details while preserving technical context."""
    if not text:
        return "Unknown context"

    result = text
    # emails
    result = re.sub(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b", "[EMAIL]", result)
    # IP addresses
    result = re.sub(r"\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b", "[IP]", result)
    # user directories
    result = re.sub(r"/Users/[^/\s]+", "/Users/[USER]", result)
    result = re.sub(r"/home/[^/\s]+", "/home/[USER]", result)
    # API keys/tokens
    result = re.sub(r"\b[A-Z0-9]{20,}\b", "[TOKEN]", result)
    # personal references
    result = re.sub(
        r"\b(user|person|someone|they)\s+(said|asked|told|mentioned)",
        "when asked",
        result,
        flags=re.IGNORECASE,
    )
    return result[:200]


def _parse_extracted_experiences(response: str) -> list[dict[str, object]]:
    """Parse JSON array of experiences from LLM response."""
    json_match = re.search(r"\[[\s\S]*\]", response)
    if not json_match:
        return []
    try:
        parsed = json.loads(json_match.group(0))
        if not isinstance(parsed, list):
            return []
        return [item for item in parsed if isinstance(item, dict)]
    except (json.JSONDecodeError, ValueError):
        return []


def _get_number_setting(runtime: IAgentRuntime, key: str, fallback: float) -> float:
    """Get a numeric setting from runtime, with fallback."""
    value = runtime.get_setting(key)
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        try:
            parsed = float(value)
            if parsed == parsed:  # check for NaN
                return parsed
        except ValueError:
            pass
    return fallback


def _coerce_int(value: object | None, default: int = 0) -> int:
    if isinstance(value, bool):
        return default
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        return int(value)
    if isinstance(value, str):
        try:
            return int(value.strip())
        except ValueError:
            return default
    return default


async def _validate_experience_evaluator(
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | None = None,
) -> bool:
    """Only run every 25 messages and only on agent messages."""
    entity_id = getattr(message, "entity_id", None)
    if str(entity_id) != str(runtime.agent_id):
        return False

    # Check cooldown - only extract experiences every 25 messages to reduce token cost
    last_extraction_key = "experience-extraction:last-message-count"
    current_count_raw = await runtime.get_cache(last_extraction_key)
    current_count = _coerce_int(current_count_raw)
    new_message_count = current_count + 1

    await runtime.set_cache(last_extraction_key, str(new_message_count))

    # Trigger extraction every 25 messages
    should_extract = new_message_count % 25 == 0

    if should_extract:
        logger.info(
            "[experienceEvaluator] Triggering experience extraction after %d messages",
            new_message_count,
        )

    return should_extract


async def _handle_experience_evaluator(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: Callable[[Content], Awaitable[list[Memory]]] | None = None,
    responses: list[Memory] | None = None,
) -> ActionResult | None:
    """Extract novel learning experiences from recent conversation."""
    _ = options, callback, responses, state  # Unused parameters

    from elizaos.features.advanced_capabilities.experience.service import ExperienceService

    experience_service = runtime.get_service(EXPERIENCE_SERVICE_TYPE)
    if not isinstance(experience_service, ExperienceService):
        logger.warning("[experienceEvaluator] Experience service not available")
        return None

    room_id = getattr(message, "room_id", None)
    recent_messages = await runtime.get_memories(
        table_name="messages",
        room_id=str(room_id) if room_id else None,
        limit=10,
    )

    if len(recent_messages) < 3:
        logger.debug("[experienceEvaluator] Not enough messages for experience extraction")
        return None

    # Combine recent messages into analysis context
    conversation_parts: list[str] = []
    for m in recent_messages:
        if hasattr(m, "content") and m.content:
            text_val = getattr(m.content, "text", None)
            if text_val:
                conversation_parts.append(str(text_val))

    conversation_context = " ".join(conversation_parts)

    # Build extraction prompt
    extraction_prompt = _EXTRACT_EXPERIENCES_TEMPLATE.replace(
        "{{conversation_context}}", conversation_context
    ).replace("{{existing_experiences}}", "None")

    # Use TEXT_SMALL -- extraction is a structured JSON task, not complex reasoning
    response = await runtime.use_model(ModelType.TEXT_SMALL, prompt=extraction_prompt)

    experiences = _parse_extracted_experiences(str(response))
    threshold = _get_number_setting(runtime, "AUTO_RECORD_THRESHOLD", 0.6)

    # Record each novel experience (max 3 per extraction)
    recorded_count = 0
    for exp in experiences[:3]:
        learning = exp.get("learning")
        confidence = exp.get("confidence")

        if not learning or not isinstance(confidence, (int, float)) or confidence < threshold:
            continue

        learning_str = str(learning)

        # Post-extraction dedup: skip if a very similar experience already exists
        similar = await experience_service.find_similar_experiences(learning_str, 1)
        if similar:
            existing_learning = similar[0].learning.lower()
            new_learning = learning_str.lower()
            existing_words = {w for w in existing_learning.split() if len(w) > 3}
            new_words = {w for w in new_learning.split() if len(w) > 3}
            overlap = len(new_words & existing_words)
            union = len(existing_words | new_words)
            if union > 0 and overlap / union > 0.6:
                logger.debug(
                    '[experienceEvaluator] Skipping duplicate experience: "%s..."',
                    learning_str[:80],
                )
                continue

        normalized_type = str(exp.get("type", "")).upper()
        experience_type = _EXPERIENCE_TYPE_MAP.get(normalized_type, ExperienceType.LEARNING)

        await experience_service.record_experience(
            {
                "type": experience_type,
                "outcome": (
                    OutcomeType.POSITIVE
                    if experience_type == ExperienceType.CORRECTION
                    else OutcomeType.NEUTRAL
                ),
                "context": _sanitize_context(str(exp.get("context", "Conversation analysis"))),
                "action": "pattern_recognition",
                "result": learning_str,
                "learning": _sanitize_context(learning_str),
                "domain": _detect_domain(learning_str),
                "tags": ["extracted", "novel", experience_type.value],
                "confidence": min(float(confidence), 0.9),  # Cap confidence
                "importance": 0.8,  # High importance for extracted experiences
            }
        )

        logger.info(
            "[experienceEvaluator] Recorded novel experience: %s...",
            learning_str[:100],
        )
        recorded_count += 1

    if recorded_count > 0:
        logger.info(
            "[experienceEvaluator] Extracted %d novel experiences from conversation",
            recorded_count,
        )
    else:
        logger.debug("[experienceEvaluator] No novel experiences found in recent conversation")

    return ActionResult(
        success=True,
        data={"extractedCount": recorded_count},
        values={"extractedCount": str(recorded_count)},
    )


experience_evaluator = Evaluator(
    name="EXPERIENCE_EVALUATOR",
    description="Periodically analyzes conversation patterns to extract novel learning experiences",
    handler=_handle_experience_evaluator,
    validate=_validate_experience_evaluator,
    similes=["experience recorder", "learning evaluator", "self-reflection"],
    always_run=False,
    examples=[],
)

File: packages/typescript/src/features/advanced-capabilities/experience/evaluators/experienceEvaluator.ts
import { logger } from "../../../../logger.ts";
import type {
	ActionResult,
	Evaluator,
	HandlerCallback,
	HandlerOptions,
} from "../../../../types/components.ts";
import type { Memory } from "../../../../types/memory.ts";
import { ModelType } from "../../../../types/model.ts";
import type { IAgentRuntime } from "../../../../types/runtime.ts";
import type { State } from "../../../../types/state.ts";
import { composePrompt } from "../../../../utils.ts";
import { EXTRACT_EXPERIENCES_TEMPLATE } from "../generated/prompts/typescript/prompts.ts";
import type { ExperienceService } from "../service";
import { ExperienceType, OutcomeType } from "../types";

type ExtractedExperience = {
	type?: string;
	learning?: string;
	context?: string;
	confidence?: number;
	reasoning?: string;
};

export const experienceEvaluator: Evaluator = {
	name: "EXPERIENCE_EVALUATOR",
	similes: ["experience recorder", "learning evaluator", "self-reflection"],
	description:
		"Periodically analyzes conversation patterns to extract novel learning experiences",
	alwaysRun: false,

	examples: [
		{
			prompt:
				"The agent successfully executed a shell command after initially failing",
			messages: [
				{
					name: "Autoliza",
					content: {
						text: "Let me try to run this Python script.",
					},
				},
				{
					name: "Autoliza",
					content: {
						text: "Error: ModuleNotFoundError for pandas. I need to install it first.",
					},
				},
				{
					name: "Autoliza",
					content: {
						text: "After installing pandas, the script ran successfully and produced the expected output.",
					},
				},
			],
			outcome:
				"Record a CORRECTION experience about needing to install dependencies before running Python scripts",
		},
		{
			prompt: "The agent discovered a new system capability",
			messages: [
				{
					name: "Autoliza",
					content: {
						text: "I found that the system has jq installed, which is perfect for parsing JSON data.",
					},
				},
			],
			outcome:
				"Record a DISCOVERY experience about the availability of jq for JSON processing",
		},
	],

	async validate(
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
	): Promise<boolean> {
		// Only run every 10 messages and only on agent messages
		if (message.entityId !== runtime.agentId) {
			return false;
		}

		// Check cooldown - only extract experiences every 25 messages to reduce token cost
		const lastExtractionKey = "experience-extraction:last-message-count";
		const currentCount =
			(await runtime.getCache<string>(lastExtractionKey)) || "0";
		const messageCount = Number.parseInt(currentCount, 10);
		const newMessageCount = messageCount + 1;

		await runtime.setCache(lastExtractionKey, newMessageCount.toString());

		// Trigger extraction every 25 messages (was 10 — reduced to cut LLM costs by ~60%)
		const shouldExtract = newMessageCount % 25 === 0;

		if (shouldExtract) {
			logger.info(
				`[experienceEvaluator] Triggering experience extraction after ${newMessageCount} messages`,
			);
		}

		return shouldExtract;
	},

	async handler(
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		_callback?: HandlerCallback,
		_responses?: Memory[],
	): Promise<ActionResult | undefined> {
		void _options;
		void _callback;
		void _responses;
		void state;

		const experienceService = runtime.getService(
			"EXPERIENCE",
		) as ExperienceService | null;

		if (!experienceService) {
			logger.warn("[experienceEvaluator] Experience service not available");
			return;
		}

		const recentMessages = await runtime.getMemories({
			tableName: "messages",
			roomId: message.roomId,
			limit: 10,
			unique: false,
		});

		if (recentMessages.length < 3) {
			logger.debug(
				"[experienceEvaluator] Not enough messages for experience extraction",
			);
			return;
		}

		// Combine recent messages into analysis context
		const conversationContext = recentMessages
			.map((m: Memory) => m.content.text)
			.filter(Boolean)
			.join(" ");

		// NOTE: We intentionally do NOT query existing experiences to embed in the prompt.
		// That added ~500-1000 tokens per call for dedup the LLM can't reliably do anyway.
		// Deduplication happens post-extraction via similarity check before recording.

		const extractionPrompt = composePrompt({
			state: {
				conversation_context: conversationContext,
				existing_experiences: "None",
			},
			template: EXTRACT_EXPERIENCES_TEMPLATE,
		});

		// Use TEXT_SMALL — extraction is a structured JSON task, not complex reasoning.
		// Saves 5-10x in token cost vs TEXT_LARGE.
		const runModel = runtime.useModel.bind(runtime);
		const response = await runModel(ModelType.TEXT_SMALL, {
			prompt: extractionPrompt,
		});

		const experiences = parseExtractedExperiences(response);

		const threshold = getNumberSetting(runtime, "AUTO_RECORD_THRESHOLD", 0.6);

		// Record each novel experience
		const experienceTypeMap: Record<string, ExperienceType> = {
			DISCOVERY: ExperienceType.DISCOVERY,
			CORRECTION: ExperienceType.CORRECTION,
			SUCCESS: ExperienceType.SUCCESS,
			LEARNING: ExperienceType.LEARNING,
		};

		for (const exp of experiences.slice(0, 3)) {
			// Max 3 experiences per extraction
			if (
				!exp.learning ||
				typeof exp.confidence !== "number" ||
				exp.confidence < threshold
			) {
				continue;
			}

			// Post-extraction dedup: skip if a very similar experience already exists
			const similar = await experienceService.findSimilarExperiences(
				exp.learning,
				1,
			);
			if (similar.length > 0) {
				// If the most similar existing experience shares a lot of the same words,
				// it's likely a duplicate — skip recording
				const existingLearning = similar[0].learning.toLowerCase();
				const newLearning = exp.learning.toLowerCase();
				const existingWords = new Set(
					existingLearning.split(/\s+/).filter((w) => w.length > 3),
				);
				const newWords = new Set(
					newLearning.split(/\s+/).filter((w) => w.length > 3),
				);
				const overlap = [...newWords].filter((w) =>
					existingWords.has(w),
				).length;
				const union = new Set([...existingWords, ...newWords]).size;
				if (union > 0 && overlap / union > 0.6) {
					logger.debug(
						`[experienceEvaluator] Skipping duplicate experience: "${exp.learning.substring(0, 80)}..."`,
					);
					continue;
				}
			}

			const normalizedType =
				typeof exp.type === "string" ? exp.type.toUpperCase() : "";
			const experienceType =
				experienceTypeMap[normalizedType] ?? ExperienceType.LEARNING;
			const experienceTag = experienceType;

			await experienceService.recordExperience({
				type: experienceType,
				outcome:
					experienceType === ExperienceType.CORRECTION
						? OutcomeType.POSITIVE
						: OutcomeType.NEUTRAL,
				context: sanitizeContext(exp.context || "Conversation analysis"),
				action: "pattern_recognition",
				result: exp.learning,
				learning: sanitizeContext(exp.learning),
				domain: detectDomain(exp.learning),
				tags: ["extracted", "novel", experienceTag],
				confidence: Math.min(exp.confidence, 0.9), // Cap confidence
				importance: 0.8, // High importance for extracted experiences
			});

			logger.info(
				`[experienceEvaluator] Recorded novel experience: ${exp.learning.substring(0, 100)}...`,
			);
		}

		if (experiences.length > 0) {
			logger.info(
				`[experienceEvaluator] Extracted ${experiences.length} novel experiences from conversation`,
			);
		} else {
			logger.debug(
				"[experienceEvaluator] No novel experiences found in recent conversation",
			);
		}

		return {
			success: true,
			data: {
				extractedCount: experiences.length,
			},
			values: {
				extractedCount: experiences.length.toString(),
			},
		};
	},
};

function parseExtractedExperiences(response: string): ExtractedExperience[] {
	const jsonMatch = response.match(/\[[\s\S]*\]/);
	if (!jsonMatch) return [];

	try {
		const parsed = JSON.parse(jsonMatch[0]) as ExtractedExperience[];
		if (!Array.isArray(parsed)) return [];
		return parsed.filter((item) => item && typeof item === "object");
	} catch {
		return [];
	}
}

function getNumberSetting(
	runtime: IAgentRuntime,
	key: string,
	fallback: number,
): number {
	const value = runtime.getSetting(key);
	if (typeof value === "number") return value;
	if (typeof value === "string") {
		const parsed = Number.parseFloat(value);
		return Number.isFinite(parsed) ? parsed : fallback;
	}
	return fallback;
}

function sanitizeContext(text: string): string {
	if (!text) return "Unknown context";

	// Remove user-specific details while preserving technical context
	return text
		.replace(/\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g, "[EMAIL]") // emails
		.replace(/\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b/g, "[IP]") // IP addresses
		.replace(/\/Users\/[^/\s]+/g, "/Users/[USER]") // user directories
		.replace(/\/home\/[^/\s]+/g, "/home/[USER]") // home directories
		.replace(/\b[A-Z0-9]{20,}\b/g, "[TOKEN]") // API keys/tokens
		.replace(
			/\b(user|person|someone|they)\s+(said|asked|told|mentioned)/gi,
			"when asked",
		) // personal references
		.substring(0, 200); // limit length
}

function detectDomain(text: string): string {
	const domains: Record<string, string[]> = {
		shell: ["command", "terminal", "bash", "shell", "execute", "script", "cli"],
		coding: [
			"code",
			"function",
			"variable",
			"syntax",
			"programming",
			"debug",
			"typescript",
			"javascript",
		],
		system: [
			"file",
			"directory",
			"process",
			"memory",
			"cpu",
			"system",
			"install",
			"package",
		],
		network: [
			"http",
			"api",
			"request",
			"response",
			"url",
			"network",
			"fetch",
			"curl",
		],
		data: ["json", "csv", "database", "query", "data", "sql", "table"],
		ai: ["model", "llm", "embedding", "prompt", "token", "inference"],
	};

	const lowerText = text.toLowerCase();

	for (const [domain, keywords] of Object.entries(domains)) {
		if (keywords.some((keyword) => lowerText.includes(keyword))) {
			return domain;
		}
	}

	return "general";
}
